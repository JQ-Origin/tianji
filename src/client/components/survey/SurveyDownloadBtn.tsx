import React, { useState } from 'react';
import {
  DialogHeader,
  DialogFooter,
  Dialog,
  DialogTrigger,
  DialogContent,
  DialogTitle,
  DialogDescription,
} from '@/components/ui/dialog';
import { Button } from '@/components/ui/button';
import { useTranslation } from '@i18next-toolkit/react';
import { LuCloudDownload } from 'react-icons/lu';
import dayjs from 'dayjs';
import { AppRouterOutput, trpc } from '@/api/trpc';
import { useCurrentWorkspaceId } from '@/store/user';
import { useEvent } from '@/hooks/useEvent';
import { clamp, omit } from 'lodash-es';
import { Progress } from '../ui/progress';
import jsonExport from 'jsonexport/dist';
import { downloadCSV } from '@/utils/dom';
import { message } from 'antd';
import { DatePicker, DatePickerRange } from '../DatePicker';

interface SurveyDownloadBtnProps {
  surveyId: string;
}
export const SurveyDownloadBtn: React.FC<SurveyDownloadBtnProps> = React.memo(
  (props) => {
    const { surveyId } = props;
    const workspaceId = useCurrentWorkspaceId();
    const [date, setDate] = useState<DatePickerRange | undefined>({
      from: dayjs().subtract(1, 'months').toDate(),
      to: dayjs().toDate(),
    });
    const { t } = useTranslation();

    const { data: info } = trpc.survey.get.useQuery({
      workspaceId,
      surveyId,
    });
    const { data: count = 1 } = trpc.survey.count.useQuery({
      workspaceId,
      surveyId,
    });
    const [downloadProgress, setDownloadProgress] = useState<number | null>(
      null
    );

    const trpcUtils = trpc.useUtils();

    const handleStart = useEvent(async () => {
      try {
        const limit = 1000;
        let cursor: string | undefined = undefined;
        const downloadResultList: AppRouterOutput['survey']['resultList']['items'][number][] =
          [];
        setDownloadProgress(0);
        const startAt = date?.from
          ? dayjs(date.from).startOf('day').valueOf()
          : undefined;
        const endAt = date?.to
          ? dayjs(date.to).endOf('day').valueOf()
          : undefined;

        while (true) {
          const { items, nextCursor } = await trpcUtils.survey.resultList.fetch(
            {
              workspaceId,
              surveyId,
              limit: 1000,
              cursor,
              startAt,
              endAt,
            }
          );

          downloadResultList.push(...items);
          setDownloadProgress(clamp(downloadResultList.length / count, 0, 1));

          cursor = nextCursor;
          if (items.length < limit) {
            // no more
            break;
          }
        }

        // download with csv file
        const fields = info?.payload.items ?? [];
        const csv = await jsonExport(
          downloadResultList.map((item) => {
            const map = fields.reduce(
              (prev, curr) => ({
                ...prev,
                [curr.label]: item.payload[curr.name],
              }),
              {} as Record<string, string>
            );

            return {
              id: item.id,
              sessionId: item.sessionId,
              ...map,
              ...omit(item, ['id', 'sessionId', 'payload']),
            };
          })
        );
        let filename = info?.name ?? surveyId;
        if (date && startAt && endAt) {
          filename += `-${dayjs(startAt).format('YYYY-MM-DD')}-${dayjs(endAt).format('YYYY-MM-DD')}`;
        }
        downloadCSV(csv, filename);
      } catch (err) {
        message.error(String(err));
      } finally {
        setDownloadProgress(null);
      }
    });

    return (
      <Dialog>
        <DialogTrigger asChild>
          <Button Icon={LuCloudDownload}>{t('Download')}</Button>
        </DialogTrigger>
        <DialogContent className="sm:max-w-[425px]">
          <DialogHeader>
            <DialogTitle>{t('Download')}</DialogTitle>
            <DialogDescription>
              {t('Download survey data with csv for further use.')}
            </DialogDescription>
          </DialogHeader>
          <div className="grid gap-4 py-4">
            <DatePicker value={date} onChange={setDate} />
          </div>
          <DialogFooter className="sm:items-center sm:justify-between">
            <div className="w-full sm:w-[120px]">
              {typeof downloadProgress === 'number' && (
                <Progress
                  value={downloadProgress * 100}
                  title={`${Number((downloadProgress * 100).toFixed(2))}%`}
                />
              )}
            </div>
            <Button
              loading={typeof downloadProgress === 'number'}
              onClick={handleStart}
            >
              {t('Start')}
            </Button>
          </DialogFooter>
        </DialogContent>
      </Dialog>
    );
  }
);
SurveyDownloadBtn.displayName = 'SurveyDownloadBtn';
