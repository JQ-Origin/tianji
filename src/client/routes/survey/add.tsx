import { createFileRoute, useNavigate } from '@tanstack/react-router';
import { useTranslation } from '@i18next-toolkit/react';
import { useEvent } from '@/hooks/useEvent';
import { useCurrentWorkspaceId } from '@/store/user';
import { defaultErrorHandler, trpc } from '@/api/trpc';
import { CommonWrapper } from '@/components/CommonWrapper';
import { routeAuthBeforeLoad } from '@/utils/route';
import {
  SurveyEditForm,
  SurveyEditFormValues,
} from '@/components/survey/SurveyEditForm';
import { ScrollArea } from '@/components/ui/scroll-area';
import { CommonHeader } from '@/components/CommonHeader';

export const Route = createFileRoute('/survey/add')({
  beforeLoad: routeAuthBeforeLoad,
  component: PageComponent,
});

function PageComponent() {
  const { t } = useTranslation();
  const workspaceId = useCurrentWorkspaceId();
  const createMutation = trpc.survey.create.useMutation({
    onError: defaultErrorHandler,
  });
  const utils = trpc.useUtils();
  const navigate = useNavigate();

  const onSubmit = useEvent(async (values: SurveyEditFormValues) => {
    const res = await createMutation.mutateAsync({
      ...values,
      workspaceId,
    });

    utils.survey.all.refetch();

    navigate({
      to: '/survey/$surveyId',
      params: {
        surveyId: res.id,
      },
    });
  });

  return (
    <CommonWrapper header={<CommonHeader title={t('Add Survey')} />}>
      <ScrollArea className="h-full overflow-hidden p-4">
        <SurveyEditForm onSubmit={onSubmit} />
      </ScrollArea>
    </CommonWrapper>
  );
}
