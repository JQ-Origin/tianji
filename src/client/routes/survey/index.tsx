import { CommonPageEmpty } from '@/components/CommonPageEmpty';
import { routeAuthBeforeLoad } from '@/utils/route';
import { useTranslation } from '@i18next-toolkit/react';
import { createFileRoute } from '@tanstack/react-router';

export const Route = createFileRoute('/survey/')({
  beforeLoad: routeAuthBeforeLoad,
  component: PageComponent,
});

function PageComponent() {
  const { t } = useTranslation();

  return (
    <CommonPageEmpty
      text={t('Not any survey has been exist')}
      buttonText={t('Add Survey Now')}
      buttonUrl="/survey/add"
    />
  );
}
