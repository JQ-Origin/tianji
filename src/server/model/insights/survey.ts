import { z } from 'zod';
import { insightsQuerySchema } from '../../utils/schema.js';
import { prisma } from '../_client.js';
import { printSQL } from '../../utils/prisma.js';
import { Prisma } from '@prisma/client';
import { FilterInfoType, FilterInfoValue, getDateArray } from '@tianji/shared';
import { get, uniq } from 'lodash-es';
import { env } from '../../utils/env.js';
import { InsightsSqlBuilder } from './shared.js';

export class SurveyInsightsSqlBuilder extends InsightsSqlBuilder {
  getTableName() {
    return 'SurveyResult';
  }

  buildSelectQueryArr() {
    const { metrics } = this.query;
    return metrics.map((item) => {
      if (item.math === 'events') {
        if (item.name === '$all_event') {
          return Prisma.sql`count(1) as "$all_event"`;
        }

        return Prisma.sql`sum(case WHEN "SurveyResult"."payload"->>'${item.name}' IS NOT NULL AND "SurveyResult"."payload"->>'${item.name}' <> '' THEN 1 ELSE 0 END) as ${Prisma.raw(`"${item.name}"`)}`;
      } else if (item.math === 'sessions') {
        if (item.name === '$all_event') {
          return Prisma.sql`count(distinct "sessionId") as "$all_event"`;
        }

        return Prisma.sql`count(distinct case WHEN "SurveyResult"."payload"->>'${item.name}' IS NOT NULL AND "SurveyResult"."payload"->>'${item.name}' <> '' THEN "sessionId" ELSE 0 END) as ${Prisma.raw(`"${item.name}"`)}`;
      }

      return null;
    });
  }

  buildGroupSelectQueryArr() {
    const { groups } = this.query;
    let groupSelectQueryArr: Prisma.Sql[] = [];
    if (groups.length > 0) {
      for (const g of groups) {
        if (!g.customGroups) {
          groupSelectQueryArr.push(
            Prisma.sql`"SurveyResult"."payload" ->> ${g.value} as "%${Prisma.raw(g.value)}"`
          );
        } else if (g.customGroups && g.customGroups.length > 0) {
          for (const cg of g.customGroups) {
            groupSelectQueryArr.push(
              Prisma.sql`${this.buildFilterQueryOperator(
                g.value,
                g.type,
                cg.filterOperator,
                cg.filterValue
              )} as "%${Prisma.raw(`${g.value}|${cg.filterOperator}|${cg.filterValue}`)}"`
            );
          }
        }
      }
    }
    return groupSelectQueryArr;
  }

  buildWhereQueryArr() {
    const { insightId, time, filters } = this.query;
    const { startAt, endAt } = time;

    return [
      // survey id
      Prisma.sql`"SurveyResult"."surveyId" = ${insightId}`,

      // date
      this.buildDateRangeQuery('"SurveyResult"."createdAt"', startAt, endAt),

      ...filters.map((filter) =>
        this.buildFilterQueryOperator(
          filter.name,
          filter.type,
          filter.operator,
          filter.value
        )
      ),
    ];
  }

  private buildFilterQueryOperator(
    name: string,
    type: FilterInfoType,
    operator: string,
    value: FilterInfoValue | null
  ): Prisma.Sql {
    const valueField = Prisma.sql`("SurveyResult"."payload"->> ${name})${type === 'number' ? Prisma.raw('::int') : type === 'boolean' ? Prisma.raw('::boolean') : Prisma.empty}`;
    return this.buildCommonFilterQueryOperator(
      type,
      operator,
      value,
      valueField
    );
  }
}

export async function insightsSurvey(
  query: z.infer<typeof insightsQuerySchema>,
  context: { timezone: string }
) {
  const { time, metrics, groups } = query;
  const { startAt, endAt, unit, timezone = context.timezone } = time;

  const builder = new SurveyInsightsSqlBuilder(query, context);
  const sql = builder.build();

  if (env.isDev) {
    printSQL(sql);
  }

  const res = await prisma.$queryRaw<{ date: string | null }[]>(sql);

  let result: {
    name: string;
    [groupName: string]: any;
    data: {
      date: string;
      value: number;
    }[];
  }[] = [];

  for (const m of metrics) {
    if (groups.length > 0) {
      for (const g of groups) {
        const allGroupValue = uniq(
          res.map((item) => get(item, `%${g.value}`) as any)
        );

        result.push(
          ...allGroupValue.map((gv) => ({
            name: m.name,
            [g.value]: gv,
            data: getDateArray(
              res
                .filter((item) => get(item, `%${g.value}`) === gv)
                .map((item) => {
                  return {
                    value: Number(get(item, m.name)),
                    date: String(item.date),
                  };
                }),
              startAt,
              endAt,
              unit,
              timezone
            ),
          }))
        );
      }
    } else {
      result.push({
        name: m.name,
        data: getDateArray(
          res.map((item) => {
            return {
              value: Number(get(item, m.name)),
              date: String(item.date),
            };
          }),
          startAt,
          endAt,
          unit,
          timezone
        ),
      });
    }
  }

  return result;
}
