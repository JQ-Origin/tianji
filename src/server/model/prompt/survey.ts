// @ts-ignore
import type { ChatCompletionMessageParam } from 'openai/resources/index.mjs';
import { prisma } from '../_client.js';
import { SurveyPayloadSchema } from '../../prisma/zod/schemas/index.js';
import { z } from 'zod';

export const classifySurveyInputSchema = z.object({
  surveyId: z.string(),
  startAt: z.number(),
  endAt: z.number(),
  runStrategy: z.enum(['skipExist', 'skipInSuggest', 'rebuildAll']),
  languageStrategy: z.enum(['default', 'user']).default('default'),
  payloadContentField: z.string(),
  suggestionCategory: z.array(z.string()),
});

export const classifySurveyMQSchema = classifySurveyInputSchema.merge(
  z.object({
    workspaceId: z.string(),
    language: z.string(),
  })
);

export const translateSurveyInputSchema = z.object({
  surveyId: z.string(),
  startAt: z.number(),
  endAt: z.number(),
  runStrategy: z.enum(['skipExist', 'rebuildAll']),
  languageStrategy: z.enum(['default', 'user']).default('user'),
  payloadContentField: z.string(),
});

export const translateSurveyMQSchema = translateSurveyInputSchema.merge(
  z.object({
    workspaceId: z.string(),
    language: z.string(),
  })
);

export async function getSurveyPrompt(
  surveyId: string
): Promise<ChatCompletionMessageParam[]> {
  const limit = 100;
  const info = await prisma.survey.findUnique({
    where: {
      id: surveyId,
    },
    select: {
      name: true,
      payload: true,
    },
  });
  const result = await prisma.surveyResult.findMany({
    where: {
      surveyId,
    },
    take: limit,
    select: {
      sessionId: true,
      payload: true,
      createdAt: true,
    },
    orderBy: {
      createdAt: 'desc',
    },
  });

  return [
    {
      role: 'system',
      content: `You are an analysis assistant for a questionnaire survey. You need to generate corresponding answers based on existing data and user questions. Please try to answer in the user's language.

The results of the questionnaire are in JSON format and are stored in the payload field. The description of the field is as follows:
${SurveyPayloadSchema.parse(info?.payload)
  .items.map((item) => `- ${item.name}: ${item.label}`)
  .join('\n')}

Here are the most recent ${limit} survey results:

${result.map((item) => JSON.stringify(item)).join('\n')}
        `,
    },
  ];
}

export const basicSurveyClassifyPromptToken = 465;

export function buildSurveyClassifyPrompt(
  data: {
    id: string;
    content: any;
  }[],
  suggestionCategory: string[],
  language: string = 'en'
) {
  return {
    prompt: `
## Role

You are a content analysis expert. You will receive a list of content from various countries in JSON format. Your job is to analyze these items and provide a classification for each record.

## Skills

- **Multilingual Proficiency**: Able to understand content in different languages.
- **Semantic Understanding**: Capable of interpreting the meaning behind the text.
- **Content Summarization & Organization**: Skilled in extracting key points and structuring them.
- **Structured Output Production**: Able to present findings in a clear, organized format.

## Action

1. Read the user's input, which is a JSON array of objects. Each object contains:
   - \`id\`: a unique identifier
   - \`content\`: text in one or more languages
2. Determine the best-fit category for each \`content\` from the list below.
3. Return a JSON object where each key is the \`id\` and the value is the chosen category.

## Example

### Input Example
\`\`\`
- {"id": "id1", "content": "This option dont appear i cant use it"}
- {"id": "id2", "content": "Toda vez que tento fazer uma persona acaba que o aplicativo saÃ­, impossibilitado de fazer uma persona"}
\`\`\`

### Output Example

\`\`\`json
{
  "id1": "Application malfunction",
  "id2": "Application crash"
}
\`\`\`

## Reference Categories

${JSON.stringify(suggestionCategory)}


## Constraints

1. **Output Format**: All responses must be returned in JSON format.
2. **Do not create new categories**: If you introduce a new category, ensure it is strictly necessary to capture the main intent of the content. If content seems to fall between two categories, pick the one that is the closest match rather than creating a new category.
3. **Key-Value Structure**: The output must be a JSON object where each \`id\` serves as the key and its corresponding classification is the value.
4. **Consistency in Data Count**: The number of entries in your output must match the number of entries in the input.
5. **Use Existing Categories**: Only create new categories if absolutely necessary.
6. **Return user perfer language**: response result should use **${language}** as response language.

`.trim(),
    question: data.map((obj) => `- ${JSON.stringify(obj)}`).join('\n'),
  };
}

export function buildSurveyTranslationPrompt(
  data: {
    id: string;
    content: any;
  }[],
  language: string = 'en'
) {
  const prompt = `
You are a translator expert.

Please help me translate those file to '${language}', direct give me json format response, strictly follow the following format and only output the following content, no other information. The NSFW format is not suitable for reading. If you need to wrap a line, you must use \n to indicate it. Do not split complete words when wrapping a line, if you need type " in a string, you must use \\"
`.trim();

  return {
    prompt,
    question: JSON.stringify(
      data.reduce((prev, curr) => {
        return {
          ...prev,
          [curr.id]: curr.content.replace(/"/g, "'"), // replace user input double quotes to single quotes to avoid json parse error
        };
      }, {})
    ),
  };
}
