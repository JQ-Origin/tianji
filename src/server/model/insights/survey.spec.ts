import { describe, expect, test } from 'vitest';
import { SurveyInsightsSqlBuilder } from './survey.js';
import { unwrapSQL } from '../../utils/prisma.js';

describe('SurveyInsightsSqlBuilder', () => {
  const insightId = 'cm658i2tqw96upkejldn8rpbs';
  const insightType = 'survey';

  test('groups', () => {
    const builder = new SurveyInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [
          {
            value: 'rating',
            type: 'string',
          },
        ],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql');
  });

  test('groups with custom bucket', () => {
    const builder = new SurveyInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [
          {
            value: 'rating',
            type: 'string',
            customGroups: [
              {
                filterOperator: 'in list',
                filterValue: ['1', '2', '3'],
              },
            ],
          },
        ],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql');
  });

  test('with filters', () => {
    const builder = new SurveyInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [
          {
            name: 'country',
            type: 'string',
            operator: 'equals',
            value: 'US',
          },
          {
            name: 'age',
            type: 'number',
            operator: 'greater than',
            value: 18,
          },
        ],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with filters');
  });

  test('different metrics', () => {
    const builder = new SurveyInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: 'rating',
            math: 'events',
          },
          {
            name: 'comment',
            math: 'events',
          },
          {
            name: '$all_event',
            math: 'sessions',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with different metrics');
  });

  test('different time units', () => {
    const builder = new SurveyInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'month',
        },
        groups: [],
      },
      {
        timezone: 'Asia/Shanghai',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with month time unit');
  });

  test('combined filters and groups', () => {
    const builder = new SurveyInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [
          {
            name: 'completed',
            type: 'boolean',
            operator: 'equals',
            value: 1,
          },
        ],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [
          {
            value: 'userType',
            type: 'string',
          },
        ],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with filters and groups');
  });

  test('date type filters', () => {
    const builder = new SurveyInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [
          {
            name: 'submittedAt',
            type: 'date',
            operator: 'between',
            value: [1739203200000, 1741881599999],
          },
        ],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with date filters');
  });

  test('multiple string filters', () => {
    const builder = new SurveyInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [
          {
            name: 'source',
            type: 'string',
            operator: 'contains',
            value: 'web',
          },
          {
            name: 'feedback',
            type: 'string',
            operator: 'not contains',
            value: 'bug',
          },
        ],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with multiple string filters');
  });

  test('numeric filters with ranges', () => {
    const builder = new SurveyInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [
          {
            name: 'score',
            type: 'number',
            operator: 'between',
            value: [7, 10],
          },
        ],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with numeric range filter');
  });
});
