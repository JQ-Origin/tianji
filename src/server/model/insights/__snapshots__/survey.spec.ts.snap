// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SurveyInsightsSqlBuilder > combined filters and groups > sql with filters and groups 1`] = `
"select
      to_char(date_trunc('day', "SurveyResult"."createdAt" at time zone 'Asia/Shanghai'), 'YYYY-MM-DD') date , "SurveyResult"."payload" ->> 'userType' as "%userType" , count(1) as "$all_event"
    from "SurveyResult" 
    where "SurveyResult"."surveyId" = 'cm658i2tqw96upkejldn8rpbs' AND "SurveyResult"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND ("SurveyResult"."payload"->> 'completed')::boolean = '1'
    group by 1 , 2"
`;

exports[`SurveyInsightsSqlBuilder > date type filters > sql with date filters 1`] = `
"select
      to_char(date_trunc('day', "SurveyResult"."createdAt" at time zone 'Asia/Shanghai'), 'YYYY-MM-DD') date , count(1) as "$all_event"
    from "SurveyResult" 
    where "SurveyResult"."surveyId" = 'cm658i2tqw96upkejldn8rpbs' AND "SurveyResult"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND ("SurveyResult"."payload"->> 'submittedAt') BETWEEN '2025-02-10T16:00:00.000Z' AND '2025-03-13T15:59:59.999Z'
    group by 1"
`;

exports[`SurveyInsightsSqlBuilder > different metrics > sql with different metrics 1`] = `
"select
      to_char(date_trunc('day', "SurveyResult"."createdAt" at time zone 'Asia/Shanghai'), 'YYYY-MM-DD') date , sum(case WHEN "SurveyResult"."payload"->>''rating'' IS NOT NULL AND "SurveyResult"."payload"->>''rating'' <> '' THEN 1 ELSE 0 END) as "rating" , sum(case WHEN "SurveyResult"."payload"->>''comment'' IS NOT NULL AND "SurveyResult"."payload"->>''comment'' <> '' THEN 1 ELSE 0 END) as "comment" , count(distinct "sessionId") as "$all_event"
    from "SurveyResult" 
    where "SurveyResult"."surveyId" = 'cm658i2tqw96upkejldn8rpbs' AND "SurveyResult"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz
    group by 1"
`;

exports[`SurveyInsightsSqlBuilder > different time units > sql with month time unit 1`] = `
"select
      to_char(date_trunc('month', "SurveyResult"."createdAt" at time zone 'Asia/Shanghai'), 'YYYY-MM-01') date , count(1) as "$all_event"
    from "SurveyResult" 
    where "SurveyResult"."surveyId" = 'cm658i2tqw96upkejldn8rpbs' AND "SurveyResult"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz
    group by 1"
`;

exports[`SurveyInsightsSqlBuilder > groups > sql 1`] = `
"select
      to_char(date_trunc('day', "SurveyResult"."createdAt" at time zone 'Asia/Shanghai'), 'YYYY-MM-DD') date , "SurveyResult"."payload" ->> 'rating' as "%rating" , count(1) as "$all_event"
    from "SurveyResult" 
    where "SurveyResult"."surveyId" = 'cm658i2tqw96upkejldn8rpbs' AND "SurveyResult"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz
    group by 1 , 2"
`;

exports[`SurveyInsightsSqlBuilder > groups with custom bucket > sql 1`] = `
"select
      to_char(date_trunc('day', "SurveyResult"."createdAt" at time zone 'Asia/Shanghai'), 'YYYY-MM-DD') date , ("SurveyResult"."payload"->> 'rating') IN ('1' , '2' , '3') as "%rating|in list|1,2,3" , count(1) as "$all_event"
    from "SurveyResult" 
    where "SurveyResult"."surveyId" = 'cm658i2tqw96upkejldn8rpbs' AND "SurveyResult"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz
    group by 1 , 2"
`;

exports[`SurveyInsightsSqlBuilder > multiple string filters > sql with multiple string filters 1`] = `
"select
      to_char(date_trunc('day', "SurveyResult"."createdAt" at time zone 'Asia/Shanghai'), 'YYYY-MM-DD') date , count(1) as "$all_event"
    from "SurveyResult" 
    where "SurveyResult"."surveyId" = 'cm658i2tqw96upkejldn8rpbs' AND "SurveyResult"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND ("SurveyResult"."payload"->> 'source') LIKE '%web%' AND ("SurveyResult"."payload"->> 'feedback') NOT LIKE '%bug%'
    group by 1"
`;

exports[`SurveyInsightsSqlBuilder > numeric filters with ranges > sql with numeric range filter 1`] = `
"select
      to_char(date_trunc('day', "SurveyResult"."createdAt" at time zone 'Asia/Shanghai'), 'YYYY-MM-DD') date , count(1) as "$all_event"
    from "SurveyResult" 
    where "SurveyResult"."surveyId" = 'cm658i2tqw96upkejldn8rpbs' AND "SurveyResult"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND ("SurveyResult"."payload"->> 'score')::int BETWEEN '7' AND '10'
    group by 1"
`;

exports[`SurveyInsightsSqlBuilder > with filters > sql with filters 1`] = `
"select
      to_char(date_trunc('day', "SurveyResult"."createdAt" at time zone 'Asia/Shanghai'), 'YYYY-MM-DD') date , count(1) as "$all_event"
    from "SurveyResult" 
    where "SurveyResult"."surveyId" = 'cm658i2tqw96upkejldn8rpbs' AND "SurveyResult"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND ("SurveyResult"."payload"->> 'country') = 'US' AND ("SurveyResult"."payload"->> 'age')::int > '18'
    group by 1"
`;
